### Calculating gross profit

Gross profit is the amount of money you generated in sales minus the cost of selling those goods (Also know as Cost of Goods Sold or COGS). COGS does not include administration or marketing costs, only costs directly related to production or providing the good or service, so Gross Profit is a good metric to see the core profitability of any product\service.

Gross profit margin is the gross profit expressed as a percentage of sales, which is valuable for analysis and decision making.

As the manager of a factory, you have been asked to check the gross profit forecast for the following year.

In [226]:
# Set the sales variable to 8000
sales = 8000

# Set the cost of goods sold (cogs) variable to 5400
cogs = 5400

# Calculate the gross profit (gross_profit)
gross_profit = sales - cogs

# Print the gross profit
print("The gross profit is {}.".format(gross_profit))

The gross profit is 2600.


### Calculating net profit

The net profit includes all the other costs (and all other income) that may incur or earn that is not directly related to the product. For example the salary of the accountant who prepares your invoices, the advertising budget, travel expenses, etc.

So, how do we calculate this?

We start by setting a list for all the items that we want to include in the net profit calculation and calling the list "other operating expenses" or "opex". To keep it simple we will not go into any tax or exceptional expenses or income. The simplified calculation is:

The gross_profit (gross profit) variable has been set for you.

In [227]:
# Create and print the opex list
admin=700
travel=250
training=130
marketing=280
insurance=150


opex = [admin, travel, training, marketing, insurance]
print(opex)

# Calculate and print net profit
net_profit = gross_profit - sum(opex)
print("The net profit is {}.".format(net_profit))

[700, 250, 130, 280, 150]
The net profit is 1090.


### Calculating sales

To calculate a sales forecast we need to know both the price we wish to sell our product and the number of units we expect to sell. In this example we will have a look at a company T-Z producing and selling T-Shirts.

The company offers its customers two options - a basic shirt or a customized shirt. A T-shirt costs 15 USD for a basic shirt, and 25 USD for a customized one.

T-Z have forecast orders for T-Shirts for delivery next month, and want to use the same sales mix of 60% basic and 40% customized. As this mix is relatively stable, we can calculate an average sales price per unit for T-Z. The variable forecast_units has been set for you.

In [228]:
forecast_units = 950

In [229]:
# Set variables units sold and sales price of the T-shirts (basic and custom)
salesprice_basic = 15
salesprice_custom = 25

# Calculate the combined sales price taking into account the sales mix
average_sales_price = (salesprice_basic * 0.6) + (salesprice_custom * 0.4)

# Calculate the total sales for next month
sales_USD = average_sales_price * forecast_units

In [230]:
# Print the total sales
print("Next month's forecast sales figure is {:.2f} USD.".format(sales_USD))

Next month's forecast sales figure is 18050.00 USD.


### Forecasting sales with a discount

Sales are often made with special mixes and discounts. Discounts are a great marketing tool and a way to move stock. Let's look again at company T-Z.

T-Z has launched a new range of T-Shirts linked to a celebrity meme for 40 USD per T-Shirt. They have excess stock of Celebshirt1 on their shelves, and in expectation of the release of Celebshirt2 in February, they have announced a 40% discount on Celebshirt1 in February.

For this exercise, prices are in USD, and unit amounts are total units sold in the respective month. The following variables have been defined for you:

sales_price = 40

units_january = 500

units_february = 700

The January sales only include sales of Celebshirt1.

The February sales include sales of Celebshirt1 and Celebshirt2, at a ratio of 45:55.

In [231]:
sales_price = 40

units_january = 500

units_february = 700

In [232]:
# Forecast the sales of January
sales_january = sales_price * units_january

# Forecast the discounted price
dsales_price = sales_price * 0.6

# Forecast the sales of February
sales_february = (sales_price * units_february * 0.55) + (dsales_price * units_february * 0.45)

# Print the forecast sales for January and February
print("The forecast sales for January and February are {} and {} USD respectively.".format(sales_january, sales_february))

The forecast sales for January and February are 20000 and 22960.0 USD respectively.


### Calculating COGS

Cost of Goods Sold (COGS) is calculated as follows:

COGS=Openingvalueofstock - closingvalueofstock + fixedcostscurrentperiod + variablecostscurrentperiod

We assume no closing or opening balances, stock is sold in the same month that it is produced.

Material costs to produce one T-shirt is 8 USD. Labor costs are 2 USD per shirt. As you only use materials and labor to make the shirt, this is known as variable costs. The costs to rent a machine that produces these shirts is 1300 USD per month, regardless of the amount of shirts produced. This is therefore an example of a fixed cost, as fixed costs in COGS are costs directly related to the product but are incurred whether you produce the products or not.

Preloaded variables are printed in the shell.

In [233]:
machine_rental = 1300
material_costs_per_unit = 8
labor_costs_per_unit = 2
units_jan = 200
units_feb = 250

In [234]:
# Set the variables for fixed costs and variable costs
fixed_costs = machine_rental  
variable_costs_per_unit = material_costs_per_unit  + labor_costs_per_unit

# Calculate the cogs for January and February
cogs_jan = ( 200 * variable_costs_per_unit) + fixed_costs
cogs_feb = (250* variable_costs_per_unit) + fixed_costs

In [235]:
cogs_jan

3300

In [236]:
cogs_feb

3800

In [237]:
# Calculate the unit cost for January and February
unit_cost_jan = cogs_jan / units_jan
unit_cost_feb = cogs_feb / units_feb 

# Print the January and February cost per unit
print("The cost per unit for January and February are {} and {} USD respectively.".format(unit_cost_jan, unit_cost_feb))

The cost per unit for January and February are 16.5 and 15.2 USD respectively.


### Calculating the break-even point

Tracking and understanding how cost per unit varies by month is critical, as the COGS influence the gross profit. At a certain point, if we do not make enough sales in the month to cover our costs, we will enter a loss situation.

How do we know we have produced enough units to sell to cover our costs? We do this by using a concept called the break-even point. The formula is as follows:

BreakevenPoint(Units)=FixedCostsPricePerUnit−VariableCostPerUnit

    In the last exercise, we calculated a unit cost of 16.5 USD for January and 15.2 USD for February. Keep this in mind when calculating the break-even point.

Predefined variables have been printed to the shell.

In [238]:
fixed_costs = 1300
variable_costs_per_unit = 10
units_jan = 200
units_feb = 250
sales_price = 16
cogs_jan = (units_jan * variable_costs_per_unit) + fixed_costs
cogs_feb = (units_feb * variable_costs_per_unit) + machine_rental

In [239]:
# Calculate the break-even point (in units) for Wizit
break_even = fixed_costs /(sales_price  - variable_costs_per_unit)

# Print the break even point in units
print("The break even point is {} units.".format(break_even))

# Forecast the gross profit for January and February
gross_profit_jan = (sales_price*units_jan ) - cogs_jan
gross_profit_feb = (sales_price*units_feb ) - cogs_feb 

# Print the gross profit for January and February
print("The gross profit for January and February are {} and {} USD respectively.".format(gross_profit_jan, gross_profit_feb))

The break even point is 216.66666666666666 units.
The gross profit for January and February are -100 and 200 USD respectively.


### Tesla income statement

The dataset for the Tesla income statement is called income_statement.

This dataset has a final column called Trailing Twelve Months (TTM), which is the most recent 12 months of data available. We will use this to calculate a 2018 forecast for Tesla.

We are only interested in the rows 'Revenue', 'Gross profit', 'Total operating expenses', and 'Net income', so we'll create a filtered income statement to only show these rows. The filtering code uses the following pattern.

dataframe[dataframe.columnname.isin(list_of_categories)]


In [240]:
# Import pandas library
import pandas as pd

In [241]:
# Import the data
income_statement = pd.read_csv('TSLA-Income-Statement.csv', )

In [242]:
income_statement

metric  2012-12  2013-12  2014-12  \
0                                       Revenue   413.00  2013.00  3198.00   
1                               Cost of revenue   383.00  1557.00  2317.00   
2                                  Gross profit    30.00   456.00   882.00   
3                            Operating expenses      NaN      NaN      NaN   
4                      Research and development   274.00   232.00   465.00   
5             Sales, General and administrative   150.00   286.00   604.00   
6                      Total operating expenses   424.00   518.00  1068.00   
7                              Operating income  -394.00   -61.00  -187.00   
8                              Interest Expense     0.00    33.00   101.00   
9                        Other income (expense)    -2.00    23.00     3.00   
10                          Income before taxes  -396.00   -71.00  -285.00   
11                   Provision for income taxes     0.00     3.00     9.00   
12        Net income from continuing operations  -396.00   -74.00  -294.00   
13                                        Other      NaN      NaN      NaN   
14                                   Net income  -396.00   -74.00  -294.00   
15  Net income available to common shareholders  -396.00   -74.00  -294.00   
16                           Earnings per share      NaN      NaN      NaN   
17                                        Basic    -3.69    -0.62    -2.36   
18                                      Diluted    -3.69    -0.62    -2.36   
19          Weighted average shares outstanding      NaN      NaN      NaN   
20                                        Basic   107.00   119.00   125.00   
21                                      Diluted   107.00   119.00   125.00   
22                                       EBITDA  -367.00    68.00    48.00   

    2015-12  2016-12       TTM  
0   4046.00  7000.00  10755.00  
1   3123.00  5401.00   8536.00  
2    924.00  1599.00   2219.00  
3       NaN      NaN       NaN  
4    718.00   834.00   1269.00  
5    922.00  1432.00   2250.00  
6   1640.00  2267.00   3520.00  
7   -717.00  -667.00  -1301.00  
8    119.00   199.00    390.00  
9    -40.00   120.00     53.00  
10  -876.00  -746.00  -1638.00  
11    13.00    27.00     52.00  
12  -889.00  -773.00  -1689.00  
13      NaN    98.00    282.00  
14  -889.00  -675.00  -1407.00  
15  -889.00  -675.00  -1407.00  
16      NaN      NaN       NaN  
17    -6.93    -4.68     -8.54  
18    -6.93    -4.68     -8.54  
19      NaN      NaN       NaN  
20   128.00   144.00    162.00  
21   128.00   144.00    162.00  
22  -334.00   400.00    246.00

In [243]:
# Choose some interesting metrics
interesting_metrics = ['Revenue', 'Gross profit', 'Total operating expenses', 'Net income']

# Filter for rows containing these metrics
filtered_income_statement = income_statement[income_statement.metric.isin(interesting_metrics)]

# See the result
print(filtered_income_statement)

                      metric  2012-12  2013-12  2014-12  2015-12  2016-12  \
0                    Revenue    413.0   2013.0   3198.0   4046.0   7000.0   
2               Gross profit     30.0    456.0    882.0    924.0   1599.0   
6   Total operating expenses    424.0    518.0   1068.0   1640.0   2267.0   
14                Net income   -396.0    -74.0   -294.0   -889.0   -675.0   

        TTM  
0   10755.0  
2    2219.0  
6    3520.0  
14  -1407.0  


### Forecasting profit for Tesla

As in the previous exercise, the dataset for the Tesla income statement is named income_statement. Using what we have learned in the previous exercise, we will now append a new column with 2018 Forecast data, which we will assign the header "Forecast".

For this exercise, we would like to set the filtered_income_statement to only show the row 'Revenue'.

Remember, the TTM column is the most recent 12-month value that we will use for the 2018 forecast. Thus far, we have the following information for 2018:

    The market demand analysis predicts the revenue to increase to 13,000 in 2018 due to increased sales of Model 3.


In [244]:
revenue_metric = ['Revenue']

# Filter for rows containing the revenue metric
filtered_income_statement = income_statement[income_statement.metric.isin(revenue_metric)]

# Get the number of columns in filtered_income_statement
n_cols = len(filtered_income_statement.columns)

# Insert a column in the correct position containing the column 'Forecast'
filtered_income_statement.insert(n_cols, 'Forecast', 13000) 

# See the result
print(filtered_income_statement)

    metric  2012-12  2013-12  2014-12  2015-12  2016-12      TTM  Forecast
0  Revenue    413.0   2013.0   3198.0   4046.0   7000.0  10755.0     13000


# Balance sheet and forecast ratios

### Introduction to the balance sheet

### Calculating accounts receivable (debtors)

When we sell something on credit, the credit portion is in the balance sheet under ‘Accounts Receivable’ or ‘Debtors’. For example, if credit sales are made in January with a 60-day payback period, they would be recorded in our ‘Debtors’ account in January, but only be paid (released) in March, and so on.

In this exercise, we will create the following lists:

    The credit sales in the month credits, which in this exercise is 60% of the sale value.
    The total accounts receivable debtors, to be calculated as the credits for the current month, plus the credits of the month before, minus the credits of two months before (as we assume the credits from 2 months ago or 60 days, will be repaid by then).

We have set an index for the variable month. The month value is set at 0.

In [245]:
month = 0

In [246]:
# Create the list for sales, and empty lists for debtors and credits
sales = [500, 350, 700]
credits = [] 
debtors = []

# Create the statement to append the calculated figures to the debtors and credits lists
for mvalue in sales: 
    credits.append(mvalue * 0.6)
    if month > 0:
        debtors.append(credits[month] + credits[month-1])
    else:
        debtors.append(credits[month]) 
    month += 1
# Print the result
print("The ‘Debtors’ are {}.".format(debtors))

The ‘Debtors’ are [300.0, 510.0, 630.0].


### Bad debts

When offering credit terms to customers, there is always a risk that the customer does not pay their debt. In the finance world, this is known as "bad debts".

As we have already recorded sales, we need to record the loss of sales now, as we never received the payment.

This affects both the income statement and the balance sheet. In the income statement, we record a negative value in the sales for the month we write off the debt. In the balance sheet, we need to reduce our debtor’s asset.

The following variables have been defined for January: debtors_jan = 1500

In February, we received news that a customer has gone into liquidation. This customer currently owes 500 USD.

We expect to recover 70% of this amount; the rest has to be written off as bad debts.

In [247]:
debtors_jan = 1500

In [248]:
# Calculate the bad debts for February
bad_debts_feb = 500 * 0.3

# Calculate the feb debtors amount
debtors_feb = (debtors_jan  - bad_debts_feb)

# Print the debtors for January and the bad debts and the debtors for February
print("The debtors are {} in January, {} in February. February's bad debts are {} USD.".format(debtors_jan, debtors_feb, bad_debts_feb))

The debtors are 1500 in January, 1350.0 in February. February's bad debts are 150.0 USD.


### Calculating accounts payable (creditors)

Now we will look at a scenario where we are the ones being granted credit. This means that we can buy something, but only have to pay for this amount later.

In this exercise, T-Z needs to buy nuts and bolts to produce 1000 units in January and 1200 units in February. The cost of nuts and bolts per unit is 0.25 USD. The credit terms are 50% cash upfront and 50% in 30 days.

Therefore, the creditors' value, in this case, would be paid the month directly after. This means that the creditors' value would only reflect the current month's credit purchases.

In [249]:
# Set the cost per unit
unit_cost = 0.25

# Create the list for production units and empty list for creditors
production = [1000,1200]
creditors = []

# Calculate the accounts payable for January and February
for mvalue in production: 
    creditors.append(mvalue * unit_cost * 0.5)
    
# Print the creditors balance for January and February
print("The creditors balance for January and February are {} and {} USD.".format(creditors[0], creditors[1]))

The creditors balance for January and February are 125.0 and 150.0 USD.


# Balance sheet efficiency ratios

### Debtor days ratio

The first ratio we will look at is debtor days. This ratio looks at how many days it takes to receive our money from our debtors. It is usually calculated over a period of 1 financial year.

The ratio is as follows:

The following information is available to you:

    Sales for the year: 12,500 USD
    Ending Debtors balance: 650


In [250]:
# Create the variables
debtors_end = 650
sales_tot = 12500

# Calculate the debtor days variable
ddays_ratio = (debtors_end/sales_tot) * 365

# Print the result
print("The debtor days ratio is {}.".format(ddays_ratio))


The debtor days ratio is 18.98.


### Days payable outstanding

W# Get the variables
cogs_tot = 4000
creditors_end = 650

# Calculate the days payable outstanding
dpo = (creditors_end/cogs_tot)*365

# Print the days payable outstanding
print("The days payable outstanding is {}.".format(dpo))e will now have a look at our accounts payable, or creditors, and a ratio called the Days Payable Outstanding (DPO).

This ratio is an efficiency ratio that measures the average number of days a company takes to pay its suppliers.

The DPO ratio is calculated as follows:

T-Z wants to know its days payable outstanding and has asked you to calculate it.

The following is available in the workspace:

    Total COGS for the period = 4000
    Closing Payables balance = 650


In [251]:
# Get the variables
cogs_tot = 4000
creditors_end = 650

# Calculate the days payable outstanding
dpo = (creditors_end/cogs_tot)*365

# Print the days payable outstanding
print("The days payable outstanding is {}.".format(dpo))

The days payable outstanding is 59.3125.


### Days in inventory and asset turnover ratio

In this exercise, we will calculate the time it takes for a company to turn inventory into sales (days in inventory or DII ratio) and a ratio to calculate the efficiency of a company’s assets by seeing how the company uses its assets to generate sales (asset turnover ratio). The required formulas are:

The following are preloaded:
Metric 	Variable 	Value
Total COGS 	cogs_tot 	4000
Average Inventory 	av_inv 	1900
Total Sales 	sales_tot 	10000
Opening balance Assets 	ob_assets 	2000
Closing balance Assets 	cb_assets 	7000

In [252]:
# Get the variables
cogs_tot = 4000
creditors_end = 650
av_inv = 1900
sales_tot = 10000
ob_assets = 2000
cb_assets = 7000

In [253]:
# Calculate the dii ratio 
dii_ratio = (av_inv/cogs_tot )*365

# Print the result
print("The DII ratio is {}.".format(dii_ratio))

The DII ratio is 173.375.


In [254]:
# Calculate the Average Assets
av_assets = (ob_assets + cb_assets)/2

# Calculate the Asset Turnover Ratio
at_ratio = sales_tot/av_assets

# Print the Asset Turnover Ratio
print("The asset turnover ratio is {}.".format(at_ratio))


The asset turnover ratio is 2.2222222222222223.


### Calculating balance sheet ratios for Ford

Now we will look at a real life example, Ford Inc, a company producing motor vehicles. We have uploaded a dataset: balance_sheet with the data for Ford Inc's most recent balance sheet. The sales and cost of sales figures have been provided for 2017 with the Key_Figures_Memodataset.

We are only interested in one line on the balance sheet, the Receivables (another name for Debtors), and therefore need to create a filter for this. In this exercise, we will use boolean indexing to filter our dataset for Receivables in the metric column. We will first specify our metric of interest ('Receivables'), and then check whether the column of interest has this value in each row. This will generate a boolean series of True and False values. With this series, we can then filter our existing dataset.

Once we have filtered our dataset, we can retrieve the receivables values from the most recent time period and calculate the debtor days ratio (provided below).

The balance_sheet and sales figures are provided.

In [255]:
# Import the data
balance_sheet = pd.read_csv('F-Balance-Sheet.csv', )

In [256]:
balance_sheet

metric   2013-12   2014-12   2015-12  \
0                                       Assets       NaN       NaN       NaN   
1                               Current assets       NaN       NaN       NaN   
2                                         Cash       NaN       NaN       NaN   
3                    Cash and cash equivalents   14468.0   10757.0   14272.0   
4                       Short-term investments   22100.0   20393.0   20904.0   
5                                   Total cash   36568.0   31150.0   35176.0   
6                                  Receivables   87309.0   92819.0  101975.0   
7                                  Inventories    7708.0    7866.0    8319.0   
8                         Other current assets       NaN       NaN       NaN   
9                         Total current assets  131585.0  131835.0  145470.0   
10                          Non-current assets       NaN       NaN       NaN   
11               Property, plant and equipment       NaN       NaN       NaN   
12         Gross property, plant and equipment   59092.0   59260.0   57629.0   
13                    Accumulated Depreciation  -31476.0  -29134.0  -27466.0   
14           Net property, plant and equipment   27616.0   30126.0   30163.0   
15                Equity and other investments   23663.0   26574.0   30317.0   
16                       Deferred income taxes   13315.0   13639.0   11509.0   
17                      Other long-term assets    5847.0    6353.0    7466.0   
18                    Total non-current assets   70441.0   76692.0   79455.0   
19                                Total assets  202026.0  208527.0  224925.0   
20        Liabilities and stockholders' equity       NaN       NaN       NaN   
21                                 Liabilities       NaN       NaN       NaN   
22                         Current liabilities       NaN       NaN       NaN   
23                             Short-term debt   38063.0   39172.0   42998.0   
24                         Accrued liabilities    2682.0    2987.0    2385.0   
25                           Deferred revenues    2817.0    3923.0    4559.0   
26                   Other current liabilities   30569.0   31059.0   31320.0   
27                   Total current liabilities   74131.0   77141.0   81262.0   
28                     Non-current liabilities       NaN       NaN       NaN   
29                              Long-term debt   76625.0   79999.0   89856.0   
30                  Deferred taxes liabilities     598.0     570.0     502.0   
31                         Accrued liabilities    2260.0       NaN       NaN   
32                           Deferred revenues    2534.0    2686.0    2833.0   
33                 Pensions and other benefits   14790.0   15712.0    9541.0   
34                           Minority interest      33.0      27.0      15.0   
35                 Other long-term liabilities    4672.0    7587.0   12274.0   
36               Total non-current liabilities  101512.0  106581.0  115021.0   
37                           Total liabilities  175643.0  183722.0  196283.0   
38                        Stockholders' equity       NaN       NaN       NaN   
39                                Common stock      40.0      40.0      41.0   
40                  Additional paid-in capital   21422.0   21089.0   21421.0   
41                           Retained earnings   23658.0   24556.0   14414.0   
42                              Treasury stock    -506.0    -848.0    -977.0   
43      Accumulated other comprehensive income  -18231.0  -20032.0   -6257.0   
44                  Total stockholders' equity   26383.0   24805.0   28642.0   
45  Total liabilities and stockholders' equity  202026.0  208527.0  224925.0   

     2016-12   2017-12  
0        NaN       NaN  
1        NaN       NaN  
2        NaN       NaN  
3    15905.0   18492.0  
4    22922.0   20435.0  
5    38827.0   38927.0  
6    57368.0   62809.0  
7     8898.0   10277.0  
8     3368.0    3889.0  
9   108461.0  115902.0  
10  

In [257]:
# Create the filter metric for Receivables
receivables_metric = ['Receivables']

# Create a boolean series with your metric
receivables_filter = balance_sheet.metric.isin(receivables_metric)

# Use the series to filter the dataset
filtered_balance_sheet = balance_sheet[receivables_filter]

In [258]:
filtered_balance_sheet


metric  2013-12  2014-12   2015-12  2016-12  2017-12
6  Receivables  87309.0  92819.0  101975.0  57368.0  62809.0

In [259]:
# Extract the zeroth value from the last time period (2017-12)
debtors_end = filtered_balance_sheet['2017-12'].iloc[0]

In [260]:
debtors_end

62809.0

In [261]:
sales = 156776

In [262]:
# Calculate the debtor days ratio
ddays = (debtors_end / sales) * 365

# Print the debtor days ratio
print("The debtor day ratio is {:.0f}. A higher debtors days ratio means it takes longer to collect cash from debtors.".format(ddays))

The debtor day ratio is 146. A higher debtors days ratio means it takes longer to collect cash from debtors.


### Forecasting the balance sheet for Ford

Now that we have the ratios for Ford, the management wishes to improve them for the 2018 financial year.

There is one ratio in particular that management would like to be improved – the debtor days ratio, which is currently sitting at 146 days, while better than Nissan (244 days), is still much higher than Toyota, the industry leader (31 days).

Management would like to bring the debtor days down to below 100 days.

These variables have already been set:

    filtered_balance_sheet contains the receivables row of the balance sheet.
    ddays = 99 is the target debtor days.
    sales = 156776 is the existing sales total.

The formula to calculate debtor days is:
.

<math xmlns="http://www.w3.org/1998/Math/MathML">
  <mstyle displaystyle="true" scriptlevel="0">
    <mfrac>
      <mtext>closing balance debtors / ddays = 99</mtext>
      <mtext>sales</mtext>
    </mfrac>
  </mstyle>
  <mo>&#x2217;</mo>
  <mn>365</mn>
</math>

In [263]:
ddays = 99
sales = 156776

In [264]:
# Calculate the forecasted sales 
f_sales = sales * 1.1

# Solve for the forecasted debtors' ending balance
f_debtors_end = ddays  * f_sales  / 365

print("If sales rise by 10% and the debtor days decrease to {:.0f} then the forecasted closing balance for debtors will be {:.0f}.".format(ddays, f_debtors_end))


If sales rise by 10% and the debtor days decrease to 99 then the forecasted closing balance for debtors will be 46775.


In [265]:
# From previous step
f_sales = sales * 1.10
f_debtors_end = ddays * f_sales / 365

# Get the number of columns in the filtered balance sheet
n_cols = len(filtered_balance_sheet.columns)

# Append a Forecast column of the forecasted debtors' end balance
filtered_balance_sheet.insert(n_cols,'Forecast', f_debtors_end)

# See the result
print(filtered_balance_sheet)

        metric  2013-12  2014-12   2015-12  2016-12  2017-12      Forecast
6  Receivables  87309.0  92819.0  101975.0  57368.0  62809.0  46775.086027


## Formatting raw data, managing dates and financial periods

### Financial periods and how to work with them

Converting quarters into months

A company has a challenge in separating data into months. It has received the following data:
Quarter 	Amount
Quarter 1 	700
Quarter 2 	650

The split across the months within each quarter is equal. Your goal is to separate this data into a list format containing the amounts per month for the first two quarters.

In [266]:
# Create a list for quarters and initialize an empty list qrtlist
quarters = [700, 650]
qrtlist = []

# Create a for loop to split the quarters into months and add to qrtlist
for qrt in quarters:
 month = round(qrt / 3, 2)
 qrtlist = qrtlist + [month,month,month]
 
# Print the result
print("The values per month for the first two quarters are {}.".format(qrtlist))

The values per month for the first two quarters are [233.33, 233.33, 233.33, 216.67, 216.67, 216.67].


### Merging months into quarters

Now we saw what to do when we wanted to split quarters into months for more detailed monthly information. But what happens when the opposite is true and we wish to combine data into fewer columns? This is typical when dealing with historical data, when monthly details may not be necessary, or when we need a highly consolidated version of the data for a report.

The key here is to create an index, and then only add to the quarter total quarter in cycles of 3, or until the length of the list. We can do this with the following code:

if index % 3 == 0 or index == len(months):

This code checks whether the index divided by three yields a remainder of 0, or if the index is at the end of the list months. Thus, in a loop it will execute the specified code every three months or when it reaches the end of the list.

The monthly sales are already provided in the code as months, containing the sales from the first two quarters, as well as the first month of Q3. Your task is to generate a new list called quarters that contains the quarterly totals from the first three months (which includes the partial total of Q3).

In [267]:
# Create a months list, as well as an index, and set the quarter to 0
months = [100, 100, 150, 250, 300, 10, 20]
quarter = 0
index = 1
quarters = []

# Create for loop for quarter, print result, and increment the index
for sales in months:
    quarter += sales
    if index % 3 == 0 or index == len(months):
        quarters.append(quarter)
        quarter = 0
    index = index + 1
    
print("The quarter totals are Q1: {}, Q2: {}, Q3: {}".format(quarters[0], quarters[1], quarters[2]))

The quarter totals are Q1: 350, Q2: 560, Q3: 20


## The datetime library

### Converting date formats - simple

Sales area A in Europe and Sales area B in Australia have different date formats.

    Sale A: 4000 on 14/02/2018
    Sale B: 3000 on 2 March 2018

If we want to consolidate or compare sales periods, we need to convert to the same date format. We can easily do this by using the datetime library and the datetime.strptime(date_string, format) method, using the following directives:

In [268]:
# Import the datetime python library
from datetime import datetime

# Create a dt_object to convert the first date and print the month result
dt_object1 = datetime.strptime('14/02/2018', '%d/%m/%Y')
print(dt_object1)

# Create a dt_object to convert the second date and print the month result
dt_object2 = datetime.strptime('2 March 2018', '%d %B %Y')
print(dt_object2)

2018-02-14 00:00:00
2018-03-02 00:00:00


### Converting date formats - explicit

Let's have a look at one of the dates from the previous exercise.

    Sale A: 4000 on 14/02/2018

We used the datetime library to identify the day d, month m, and year y which could help us to identify data from datasets with different date formats. However, what about a scenario where we want to convert date formats into a specific format?

In this exercise you will convert Sale A from the format 14/02/2018 to the same date format as Sale B (i.e. 14 February 2018).

We can do this easily with built-in Python functions. Remember, to split a string we can use the .split()method.

The input for this exercise will be the datetime of Sale A.

In [269]:
# Set the variable for the datetime to convert
dt = '14/02/2018'

# Create the dictionary for the month values
mm = {'01': 'January', '02': 'February', '03': 'March'}

# Split the dt string into the different parts
day, month, year = dt.split('/')

# Print the concatenated date string
print(day + ' ' + mm[month] + ' ' + year)

14 February 2018


### Working with datasets - month totals

In this exercise, we will be exploring a dataset that has multiple sales in one month. We will create a script that will enable us to identify dates within the same month, and combine them into a new month total, and append this to the table.

We will be using the dataset df, which represents data from one of our sales areas. Print it out in the console to have a look at the data. As you can see, there were two sales in March. We will combine these sales into a single month total. We can iterate over the dataset using the .iteritems() method.

We will also be using the .split() method.

In [270]:
# Set the index to start at 0
index = 0

# Create the dictionary for the months
tt = {'Jan': 0, 'Feb': 0, 'Mar': 0 }

In [271]:
tt

{'Jan': 0, 'Feb': 0, 'Mar': 0}

In [272]:
data = {'Description':  ['Sales'],
        '14-Feb': ['3000'],
        '19-Mar': ['1200'],
        '22-Mar': ['1500']}

df = pd.DataFrame (data, columns = ['Description', '14-Feb', '19-Mar', '22-Mar'])

print (df)

  Description 14-Feb 19-Mar 22-Mar
0       Sales   3000   1200   1500


In [273]:
# Create a for loop that will iterate the date and amount values in the dataset
for date, amount in df.iteritems():
    # Create the if statement to split the day and month, then add it to the new tt variable
    if index > 0: 
        day, month = date.split('-')
        tt[month] +=float(amount[0])
    index += 1

print(tt)

{'Jan': 0, 'Feb': 3000.0, 'Mar': 2700.0}


### Working with datasets - combining datasets

In this example, we will be working with two datasets, df1 and df2. You can print these both out in the console to see what they contain. You will notice that they contain different date formatting.

More specifically, df1 specifies the month by the name (e.g. 02-Feb-18), whereas df2 specifies the month numerically (e.g. 06/01/2018). Additionally, df1 uses a hyphen (-) as a separator, whereas df2 uses a forward slash (/) as a separator.

We will be combining these two datasets to form a consolidated forecast for the quarter. To do this, we will need to parse the different date formats of df1 and df2.

In [274]:
df1 = pd.DataFrame()

In [275]:
totals = {'Jan': 0, 'Feb': 0, 'Mar': 0}
calendar = {'01': 'Jan', '02': 'Feb', '03': 'Mar'}

In [276]:
totals = {'Jan': 0, 'Feb': 0, 'Mar': 0}
calendar = {'01': 'Jan', '02': 'Feb', '03': 'Mar'}

for date, amount in df1.iteritems():
        day, month, year = date.split('-')
        totals[month] +=float(amount[0]) 

In [277]:
df1

Empty DataFrame
Columns: []
Index: []

In [278]:
df2 = pd.DataFrame()

In [279]:
for date, amount in df2.iteritems():
        day, month, year = date.split('/')
        totals[calendar[month]] += float(amount[0])

print(totals)

{'Jan': 0, 'Feb': 0, 'Mar': 0}


In [280]:
df2

Empty DataFrame
Columns: []
Index: []

## Assumptions and variances in forecasts

In this chapter, we will be exploring two more aspects to creating a good forecast. First, we will look at assumptions, what drives them and what happens when an assumption changes? Next, we will look at variances, as a forecast is built at one point in time, but what happens when the actual results do not correspond to our forecast? We need to build a sensitive forecast that can be sensitive to changes in both assumptions and take into account variances, and this is what we will explore in this chapter.

### Building sensitive forecast models

### Weighted probability

Txs Tools, a company selling hardware tools, is looking to expand out of their home market A into Market B. They have done some market research, and have received the following numeric probabilities:
Sales Level (USD) 	Probability (%)




200 	10
300 	40
500 	20
800 	25

Txs Tools will only be motivated to expand if they can have reasonable assurance that they will achieve sales of 400 or more. To manage the different forecast sales probabilities, Txs Tools have asked you to calculate the weighted probability.

In [281]:
# Create the combined list for sales and probability
sales_probability = ['0|0.05', '200|0.10', '300|0.40', '500|0.20', '800|0.25'] 
weighted_probability = 0

# Create a for loop to calculate the weighted probability
for pair in sales_probability:
    parts = pair.split('|')
    weighted_probability += float(parts[0]) * float(parts[1])

# Print the weighted probability result
print("The weighted probability is {}.".format(weighted_probability))

The weighted probability is 440.0.


###  Market sentiment

Txs Tools has forecast sales of 500 in January, with an expected increase of 5% per month for the rest of the quarter.

However, this is dependent on the market sentiment. Based on historical trends, the following information has been provided:

    If the market sentiment drops below 0.6 then the sales will only be realized at an increase of 2% per month.
    If market sentiment increases above 0.8. then sales are expected to increase by 7%.


In [282]:
# Create the computevariance function
def computevariance(amount, sentiment):
 if (sentiment < 0.6):
  res = amount + (amount * 0.02)
 elif (sentiment > 0.8):
  res = amount + (amount * 0.07)
 else:
  res = amount + (amount * 0.05)
 return res

In [283]:
amount=500

In [284]:
amount

500

In [285]:
# Compute the variance for jan, feb and mar
jan = computevariance(500, 0.5)
feb = computevariance(500, 0.65)
mar = computevariance(500, 0.85)

print("The forecast sales considering variance due to market sentiment is {} for Jan, {} for Feb, and {} for Mar.".format(jan, feb, mar))

The forecast sales considering variance due to market sentiment is 510.0 for Jan, 525.0 for Feb, and 535.0 for Mar.


### Assigning dependencies for sales and COGS

xs Tools have built a monthly forecast for their gross profit. This will rely on dependencies for Sales and COGS.

Set the dependencies for sales and cogs based on the information below:

    Sales dependency sales_dep: The sale price is the net price after 1 USD commission. Commissions paid increase from 1 USD per unit to 2 USD per unit for every unit above 350 units sold.
    Cost dependency cost_dep: When sales per unit increase above 500 units, an additional production line needs to be used, causing an increase in the cost per unit above 500 of 2 USD per unit.

The baseline sale price per unit (base_sales_price) is 15 USD and the baseline cost per unit (base_cost_price) is 7 USD.

In [286]:
base_sales_price=15

In [287]:
sales=750

In [288]:
# Set the Sales Dependency
if sales >= 350:
    sales_dep = (350 * base_sales_price) + ((sales - 350) * (base_sales_price - 1))
else:
    sales_dep = sales * base_sales_price

# Print the results
print("The sales dependency is {} USD.".format(sales_dep))

The sales dependency is 10850 USD.


In [289]:
base_cost_price=7

In [290]:
# Set the Cost Dependency
if sales >= 500:
    cost_dep = (500 * base_cost_price) + ((sales - 500) * (base_cost_price + 2))
else:
    cost_dep = sales * base_cost_price
    
# Print the results
print("The cost dependency is {} USD.".format(cost_dep))

The cost dependency is 5750 USD.


### Building a sensitivity analysis for gross profit

Txs Tools is now ready to use these dependencies in the gross profit forecast.

The following forecast unit sales have been provided:

Jul = 700 Aug = 350 Sep = 650

The dependencies for sales and cogs are based on the following:

    Sales dependency sales_dep: The sale price is the net price after 1 USD commission. Commissions paid increase from 1 USD per unit to 2 USD per unit for every unit above 350 units sold.
    Cost dependency cost_dep: When sales per unit increase above 500 units, an additional production line needs to be used, causing an increase in the cost per unit above 500 of 2 USD per unit.

The basic cost price base_cost_price = 7 and basic sales price base_sales_price = 15 has already been set.

In [291]:
# Create the sales_usd list
sales_usd = [700, 350, 650]

In [292]:
# Create the sales_usd list
sales_usd = [700, 350, 650]

# Create the if statement to calculate the forecast_gross_profit
for sales in sales_usd:
    if sales > 350:
        sales_dep = (350 * base_sales_price) + ((sales - 350) * (base_sales_price - 1))
    else:
        sales_dep = sales * base_sales_price
    if sales > 500:
        cost_dep = (500 * base_cost_price) + ((sales - 500) * (base_cost_price + 2))
    else:
        cost_dep = sales * base_cost_price
    forecast_gross_profit = sales_dep - cost_dep
    # Print the result
    print("The gross profit forecast for a sale unit value of {} is {} USD.".format(sales, forecast_gross_profit))
    

The gross profit forecast for a sale unit value of 700 is 4850 USD.
The gross profit forecast for a sale unit value of 350 is 2800 USD.
The gross profit forecast for a sale unit value of 650 is 4600 USD.


### Assigning dependencies for expenses

Txs Tools wants to assign a dependency for its operating expenses, particularly admin salaries.

The conditions are as follows:

    Admin expenses increase in July and August (Jul and Aug) as temporary workers need to be hired to cover the summer holiday.
    The increase is based on the number of employees taking holidays during that time. For the current year, the value for August is emp_leave = 6 (6 employees expected to take leave).
    The cost is 80 USD per temp employee hired.


In [293]:
emp_leave = 6

In [294]:
# Set the admin dependency
if emp_leave > 0:
    admin_dep = emp_leave  * 80

# Print the results
print("The admin dependency for August is {} USD.".format(admin_dep))

The admin dependency for August is 480 USD.


### Build a sensitivity analysis for the net profit

Txs Tools has provided the following forecast admin cost in USD based on full-time employees:

Jul = 1500 Aug = 1500 Sep = 1500

Build the forecast net profit forecast_net_profit when emp_leave = [6, 6, 0] and the cost per temp employee is 80 USD.

In addition to emp_leave and admin_usd, forecast gross profit has already been provided for you as forecast_gross_profit.

In [295]:
admin_usd = [1500, 1500, 1500]
emp_leave = [6, 6, 0]
forecast_gross_profit = [4850, 2800, 4600]
index = 0

for admin in admin_usd:
    temp = emp_leave[index]
    if temp > 0:
        admin_dep = emp_leave[index]* 80 + admin
    else: 
         admin_dep = admin 
    forecast_net_profit = forecast_gross_profit[index] - admin_dep
    print(forecast_net_profit)
    index += 1
    print("The forecast net profit is:")


2870
The forecast net profit is:
820
The forecast net profit is:
3100
The forecast net profit is:


## Working with variances in the forecast

### Building an alternate forecast

We will now build an alternative forecast for Txs Tools. The new quarter forecast is based off actual data for Jul - Aug as well as adjusted forecast data for September. The data (units sold) is as follows:

    Jul = 700
    Aug = 220
    Sep = 520

The dependencies calculations have already been completed from the previous exercise. The following information applies:

    base_cost_price = 7
    base_sales_price = 15


In [296]:
base_cost_price = 7
base_sales_price = 15

In [297]:
def dependencies(base_cost_price, base_sales_price, sales_usd):
    res = []
    for sales in sales_usd:
        if sales >= 350:
            sales_dep = (350 * base_sales_price) + ((sales - 350) * (base_sales_price - 1))
        else:
            sales_dep = sales * base_sales_price
        if sales >= 500:
            cost_dep = (500 * base_cost_price) + ((sales - 500) * (base_cost_price + 2))
        else:
            cost_dep = sales * base_cost_price
        res.append(sales_dep - cost_dep)
    return res

In [298]:
forecast1 = dependencies(7, 15, [700, 350, 650])

In [299]:
forecast2 = dependencies(7, 15, [700, 220, 520])

print("The original forecast scenario is {}:".format(forecast1))
print("The alternative forecast scenario is {}:".format(forecast2))

The original forecast scenario is [4850, 2800, 4600]:
The alternative forecast scenario is [4850, 1760, 3950]:


### Building a gap analysis between forecasts

Txs Tools now has two forecasts, the original forecast forecast1 and the adjusted forecast forecast2.

The dependencies have already been defined as def dependencies(base_cost_price, base_sales_price, sales_usd), where base_cost_price = 7 and base_sales_price = 15, with forecast2 based off the following adjusted sales unit values:

    Jul = 700
    Aug = 220
    Sep = 520

In this exercise, we will look at how to use a for loop to cycle between two different lists, forecast1 and forecast2 and calculate the difference ("gap") using an incremented index. It is possible to do this simultaneously as both lists have the same length.

In [300]:
# Set the two results
forecast1 = dependencies(7, 15, [700, 350, 650])
forecast2 = dependencies(7, 15, [700, 220, 520])

# Create an index and the gap analysis for the forecast
index = 0
for value in forecast2:
    print("The gap between forecasts is {}".format(value - forecast1[index]))
    index += 1

The gap between forecasts is 0
The gap between forecasts is -1040
The gap between forecasts is -650


### Setting dependencies for Netflix

Netflix has compiled a forecast up to the 2019 financial year netflix_f_is, and has based the sales figures in 2019 on the following dependency:

    Number of active subscriptions, which are based on the success of Netflix original shows.

For 2019, the success of original shows (critical and commercial acclaim) are estimated at 78%. The total amount of subscribers per percentage point is 500, and set to the variable n_subscribers_per_pp (i.e there is a calculated correlation between show success and number of subscribers).

In this exercise, we will calculate how dependent sales are on the number of subscribers in the forecast, which we will use in the next exercise.

The Netflix forecast, netflix_f_iscan be printed in the shell.

In [301]:
# Import the data
netflix_f_is = pd.read_csv('Netflix.csv', )

In [302]:
netflix_f_is

metric  2014 act  2015 act  2016 act  2017 fc  2018 fc  \
0                    Sales    5505.0    6780.0    8831.0  11688.0  14979.0   
1                   EBITDA     528.0     493.0     611.0   1088.0   1899.0   
2  Operating profit (EBIT)     403.0     306.0     380.0    837.0   1660.0   
3               Net income     267.0     123.0     187.0    559.0   1024.0   

   2019 fc  
0  17994.0  
1   2943.0  
2   2702.0  
3   1721.0

In [303]:
# Create a filter to select the sales row from the netflix_f_is dataset
sales_metric = ['Sales']

# Filter for rows containing the Sales metric
filtered_netflix_f_is = netflix_f_is[netflix_f_is.metric.isin(sales_metric)]

# Extract the 2019 Sales forecast value
forecast1 = netflix_f_is['2019 fc'].iloc[0]

# Print the resulting forecast
print("The sales forecast is {}.".format(forecast1))

The sales forecast is 17994.0.


In [304]:
n_subscribers_per_pp=500

In [305]:
# Set the success percentage to 78%
pct_success = 0.78

# Calculate the dependency for the subscriber base
n_subscribers = n_subscribers_per_pp * pct_success

# See the result
print("The dependency for the subscriber base is {}.".format(n_subscribers))

The dependency for the subscriber base is 390.0.


Calculate the ratio between forecast 2019 sales (forecast1 and the number of subscribers n_subscribers, assigning the result to sales_subs_ratio.

In [306]:
# From previous steps
sales_metric = ['Sales']
filtered_netflix_f_is = netflix_f_is[netflix_f_is.metric.isin(sales_metric)]
forecast1 = filtered_netflix_f_is["2019 fc"].iloc[0]
pct_success = 0.78
n_subscribers = n_subscribers_per_pp * pct_success

# Calculate the ratio between forecast sales and subscribers
sales_subs_ratio = forecast1 / n_subscribers

# See the result
print("The ratio between subscribers and sales is 1 subscriber equals ${:.2f}.".format(sales_subs_ratio))

The ratio between subscribers and sales is 1 subscriber equals $46.14.


### Calculating an alternative forecast for Netflix

The original assumptions are as follows: the total amount of subscribers at a 78% success rate results in 39,000 subscribers. We used this to build the forecast numbers.

However, the success rate for 2019 has been recalculated to have a probability of 65%, and the management has asked us to make an adjusted forecast based on this value.

The ratio between the subscribers and sales is 1 subscriber to 0.46 USD sales, set to variable sales_subs_ratio. In addition, the variables for number of subscribers per percentage point (n_subscribers_per_pp) and our prior calculation of forecast1 have also been set.

In [307]:
# Set the proportion of successes to 65%
pct_success2 = 0.65

# Calculate the number of subscribers
n_subscribers2 = n_subscribers_per_pp * pct_success2

# Calculate the new forecast
forecast2 = n_subscribers2 * sales_subs_ratio

In [308]:
forecast2

14995.000000000002

In [309]:
# From previous step
pct_success2 = 0.65
n_subscribers2 = n_subscribers_per_pp * pct_success2
forecast2 = n_subscribers2 * sales_subs_ratio

# Insert a column named AltForecast, containing forecast2
filtered_netflix_f_is.insert(len(filtered_netflix_f_is.columns), 'AltForecast', forecast2)

# Insert a column named Gap, containing the difference
filtered_netflix_f_is.insert(len(filtered_netflix_f_is.columns), 'Gap', (forecast1-forecast2))

# See the result
print(filtered_netflix_f_is)

  metric  2014 act  2015 act  2016 act  2017 fc  2018 fc  2019 fc  \
0  Sales    5505.0    6780.0    8831.0  11688.0  14979.0  17994.0   

   AltForecast     Gap  
0      14995.0  2999.0  
